src/scripts/main.js

```diff
@@ -1,3 +1,225 @@
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+    [0, 0, 0, 0],
+  ];
+
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+
+  setNewNumber();
+  setNewNumber();
+}
+
+// setting new 2 or 4 in random empty cells
+function setNewNumber() {
+  if (!hasEmptyCell()) {
+    return;
+  }
+
+  let found = false;
+
+  while (!found) {
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (gameField[r][c] === 0) {
+      gameField[r][c] = randomizeNumber();
+      updateCell(field.rows[r].cells[c], gameField[r][c]);
+      found = true;
+    }
+  }
+}
+
+function hasEmptyCell() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (gameField[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+// 10% probability of the appearance of 4
+function randomizeNumber() {
+  return Math.random() >= 0.9 ? 4 : 2;
+}
+
+// updating cell value
+function updateCell(fieldCell, num) {
+  fieldCell.innerText = '';
+  fieldCell.className = 'field-cell';
+  fieldCell.classList.add(`field-cell--${num}`);
+
+  if (num > 0) {
+    fieldCell.innerText = num;
+  } else {
+    fieldCell.innerText = '';
+  }
+
+  if (!fieldCell.innerText) {
+    fieldCell.className = 'field-cell';
+  }
+
+  setGameOver();
+}
+
+// cells sliding
+function slide(row) {
+  let newRow = row;
+
+  newRow = filterZero(newRow);
+
+  for (let i = 0; i < newRow.length - 1; i++) {
+    if (newRow[i] === newRow[i + 1]) {
+      newRow[i] *= 2;
+      newRow[i + 1] = 0;
+      score += newRow[i];
+
+      if (newRow[i] === 2048) {
+        winMessage.classList.remove('hidden');
+      }
+    }
+  }
+
+  newRow = filterZero(newRow);
+
+  while (newRow.length < columns) {
+    newRow.push(0);
+  }
+
+  return newRow;
+}
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slideHorizontally(side) {
+  for (let r = 0; r < rows; r++) {
+    let row = gameField[r];
+
+    if (side === 'ArrowLeft') {
+      row = slide(row);
+    }
+
+    if (side === 'ArrowRight') {
+      row.reverse();
+      row = slide(row);
+      row.reverse();
+    }
+
+    gameField[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
+
+function slideVertically(side) {
+  for (let c = 0; c < columns; c++) {
+    let row = [
+      gameField[0][c],
+      gameField[1][c],
+      gameField[2][c],
+      gameField[3][c],
+    ];
+
+    if (side === 'ArrowUp') {
+      row = slide(row);
+    }
+
+    if (side === 'ArrowDown') {
+      row.reverse();
+      row = slide(row);
+      row.reverse();
+    }
+
+    for (let r = 0; r < rows; r++) {
+      gameField[r][c] = row[r];
+
+      const num = gameField[r][c];
+
+      updateCell(field.rows[r].cells[c], num);
+    }
+  }
+}
```

Comment:
It is better, but you still repeat the code in these two functions. think about the `transpose` function and merging this logic

Reply:
Thank you!